### PR TITLE
RELATED: TNT-1502 Execution result and column totals and totals of totals

### PR DESCRIPTION
**Pull request by NestorEncinas**
---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13448/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15112/ : SUCCESS in 16m 11s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13453/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15116/ : FAILURE in 9m 48s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13514/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15174/ : FAILURE in 8m 50s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13532/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15192/ : FAILURE in 9m 13s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13534/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15194/ : FAILURE in 17m 29s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13541/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15199/ : FAILURE in 23m 40s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13542/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15200/ : FAILURE in 9m 12s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-ui/src/base/results/internal/resultDataMethods.ts [hunk 1]
```diff
@@ -122,6 +151,38 @@ class ResultDataMethods implements IResultDataMethods {
...
+    public hasRowTotals(): boolean {
+        return this.dataView.totals?.[0] !== undefined && this.dataView.totals[0].length > 0;
+    }
+
+    /**
+     *
+     * @returns - existance of totals for columns
+     */
+    public hasColumnTotals(): boolean {
+        return this.dataView.totals?.[1] !== undefined && this.dataView.totals[1].length > 0;
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui/src/base/results/internal/resultDataMethods.ts:179]: what about to reuse this.columnTotals() internally?

File: libs/sdk-ui-pivot/src/impl/data/rowFactory.ts [hunk 2]
```diff
@@ -132,15 +132,15 @@ export function getRowTotals(
...
     }
 
-    const totals = dv.rawData().totals();
+    const rowTotals = dv.rawData().rowTotals();
 
-    invariant(totals && totals.length > 0);
+    invariant(rowTotals && rowTotals.length > 0);
 
-    const colGrandTotals = totals[0];
+    const colGrandTotals = rowTotals;
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-pivot/src/impl/data/rowFactory.ts:143]: col vs row naming is confusing :(

**Review by ivan-nejezchleb — CHANGES_REQUESTED** [review]

File: libs/sdk-ui/src/base/results/internal/resultDataMethods.ts [hunk 3]
```diff
@@ -122,6 +151,38 @@ class ResultDataMethods implements IResultDataMethods {
...
+
+    public totalOfTotals(): DataValue[][][] | undefined {
+        return this.dataView.totalTotals;
+    }
+
+    /**
+     *
+     * @returns - existance of totals for rows
+     */
+    public hasRowTotals(): boolean {
```
[the remark this answers is not in the document]
Comment (pbenes) [reply libs/sdk-ui/src/base/results/internal/resultDataMethods.ts:170]: typescript validation as discussed

File: libs/sdk-ui/src/base/results/internal/resultDataMethods.ts (hunk 1 above)
Comment (pbenes) [reply libs/sdk-ui/src/base/results/internal/resultDataMethods.ts:179]: typescript validation as discussed

File: libs/sdk-backend-tiger/src/convertors/fromBackend/afm/DimensionHeaderConverter.ts [hunk 4]
```diff
@@ -81,18 +83,48 @@ function getMeasuresFromDimensions(dimensions: IDimensionDescriptor[]): IMeasure
 export function getTransformDimensionHeaders(
     dimensions: IDimensionDescriptor[],
     dateFormatter: DateFormatter,
+    grandTotals: ExecutionResultGrandTotal[],
```
[the remark this answers is not in the document]
Comment (pbenes) [reply libs/sdk-backend-tiger/src/convertors/fromBackend/afm/DimensionHeaderConverter.ts]: does not need to be, added default value, see the commits

File: libs/sdk-backend-tiger/src/convertors/fromBackend/afm/DimensionHeaderConverter.ts [hunk 5]
```diff
@@ -81,18 +83,48 @@ function getMeasuresFromDimensions(dimensions: IDimensionDescriptor[]): IMeasure
...
+            const innerTotalIndices: { [key: number]: TotalExecutionResultHeader } = {};
             return dimensionHeader.headerGroups.map((headerGroup, headerGroupIndex) => {
+                let appendedHeader: IResultTotalHeader[] = [];
+                if (dimensionIndex === 1 && columnGrandTotals) {
+                    // Append appropriate column grand total headers to each row. The result is always a total header.
+                    // This header was previously a measure header, but is converted to a total header, preserving the
+                    // linkage from the total header to the original measure with measureIndex so that formatting can work properly.
+                    // (for linkage of totals, the order of total items is used and stored in innerTotalIndices)
+                    const columnGrandTotalHeaderGroups = columnGrandTotals.dimensionHeaders[0].headerGroups;
+                    appendedHeader = columnGrandTotalHeaderGroups[headerGroupIndex].headers
```
[the remark this answers is not in the document]
Comment (pbenes) [reply libs/sdk-backend-tiger/src/convertors/fromBackend/afm/DimensionHeaderConverter.ts]: fixed

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts [hunk 6]
```diff
@@ -121,6 +125,26 @@ export function convertTotals(def: IExecutionDefinition): Total[] {
...
                         hasRowAndColumnSubTotals = true;
                     }
+
+                    // Check for rows subtotals within columns grand totals
+                    if (
+                        attributeTotal.attributeIdentifier !== attributeIdentifiers[0] &&
+                        columnTotal.attributeIdentifier === columnIdentifiers[0]
+                    ) {
+                        hasRowSubtotalAndColumnGrandTotal = true;
+                        attributeSubtotalDimensionIndex = attributeDimensionIndex =
```
[the remark this answers is not in the document]
Comment (pbenes) [reply libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts]: good catch! removed

File: libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts [hunk 7]
```diff
@@ -156,6 +180,61 @@ export function convertTotals(def: IExecutionDefinition): Total[] {
...
+             */
+            if (hasRowSubtotalAndColumnGrandTotal) {
+                const grandTotalOfSubTotal: Total[] = [
+                    {
+                        function: convertTotalType(attributeTotal.type),
+                        metric: attributeTotal.measureIdentifier,
+                        localIdentifier: subTotalColumnLocalIdentifier(attributeTotal, index),
+                        totalDimensions: [
+                            {
+                                dimensionIdentifier: "dim_1",
```
[the remark this answers is not in the document]
Comment (pbenes) [reply libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts]: switched

File: libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts [hunk 8]
```diff
@@ -217,6 +296,14 @@ export function grandTotalLocalIdentifier(total: ITotal, dimIdx: number): string
     return `total_of_totals_${total.type}_${total.measureIdentifier}_by_${total.attributeIdentifier}_${dimIdx}`;
 }
 
+export function subTotalRowLocalIdentifier(total: ITotal, dimIdx: number): string {
+    return `subtotal_row_${total.type}_${total.measureIdentifier}_by_${total.attributeIdentifier}_${dimIdx}`;
```
[the remark this answers is not in the document]
Comment (pbenes) [reply libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts:302]: Maybe @NestorEncinas would know, but probably to keep it consistent with the base one where it's also used https://github.com/gooddata/gooddata-ui-sdk/pull/3753/files/042b7ba06cfa0cb4ed5ed5e7e80065763fcf4db9#diff-dba9d8f2252d81dac2d3d12606dfaaea5ec50c27c4fc56b15222b6a6ccdab270L213 (you mean that the attributeIdentifier implies the dimension as the identifiers are different even if the same attribute is put to rows and columns, right?)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13556/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15213/ : FAILURE in 16m 59s

File: libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (NestorEncinas) [reply libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts:302]: Was to keep similar structure. And also is there as some extra check to be sure that id's does not match and make the execute fail.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13566/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15220/ : SUCCESS in 16m 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13569/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15223/ : SUCCESS in 16m 19s

**Review by ivan-nejezchleb — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3753

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by pbenes** [comment]
extended test - cypress - isolated

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3753

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2727/ : SUCCESS in 11m 40s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15251/ : SUCCESS in 16m 27s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/261/ : SUCCESS in 11m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3884/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/609/ : FAILURE in 1m 11s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/231/ : SUCCESS in 9m 02s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/118/ : SUCCESS in 4m 37s (non-voting)

**Pull request merged**